Localization/cubature_kalman_filter/cubature_kalman_filter.py

```diff
@@ -0,0 +1,234 @@
+               [0.0, 0.0, 0.0, 0.0, 1.0]])
+
+
+# r matrix - measurement noise covariance
+r = np.array([[0.015, 0.0, 0.0, 0.0],
+              [0.0, 0.010, 0.0, 0.0],
+              [0.0, 0.0, 0.1, 0.0],
+              [0.0, 0.0, 0.0, 0.01]])**2
+
+
+# main program
+def main():
+    print(__file__ + " start!!")
+    x_est = x_0
+    p_est = p_0
+    x_true = x_0
+    x_true_cat = np.array([x_0[0, 0], x_0[1, 0]])
+    x_est_cat = np.array([x_0[0, 0], x_0[1, 0]])
+    z_cat = np.array([x_0[0, 0], x_0[1, 0]])
+    for i in range(N):
+        # generate ground truth
+        x_true = f(x_true)
+        # generate measurement
+        z = gen_measurement(x_true)
+        if i == (N - 1) and show_final == 1:
+            show_final_flag = 1
+        else:
+            show_final_flag = 0
+        if show_animation == 1:
+            plot_animation(i, x_true_cat, x_est_cat, z)
+        if show_ellipse == 1:
+            plot_ellipse(x_est[0:2], p_est)
+        if show_final_flag == 1:
+            plot_final(x_true_cat, x_est_cat, z_cat)
+        x_est, p_est = cubature_kalman_filter(x_est, p_est, z)
+        x_true_cat = np.vstack((x_true_cat, np.transpose(x_true[0:2])))
+        x_est_cat = np.vstack((x_est_cat, np.transpose(x_est[0:2])))
+        z_cat = np.vstack((z_cat, np.transpose(z[0:2])))
+    print('CKF Over')
+
+
+# cubature kalman filter
+def cubature_kalman_filter(x_est, p_est, z):
+    x_pred, p_pred = cubature_prediction(x_est, p_est)
+    x_upd, p_upd = cubature_update(x_pred, p_pred, z)
+    return x_upd.astype(float), p_upd.astype(float)
+
+
+# CTRV motion model f matrix
+def f(x):
+    x[0] = x[0] + (x[3]/x[4]) * (np.sin(x[4] * dt + x[2]) - np.sin(x[2]))
+    x[1] = x[1] + (x[3]/x[4]) * (- np.cos(x[4] * dt + x[2]) + np.cos(x[2]))
+    x[2] = x[2] + x[4] * dt
+    x[3] = x[3]
+    x[4] = x[4]
+    x.reshape((5, 1))
+    return x.astype(float)
+
+
+# CTRV measurement model h matrix
+def h(x):
+    x = hx @ x
+    x.reshape((4, 1))
+    return x
+
+
+# generate sigma points
+def sigma(x, p):
+    n = np.shape(x)[0]
+    SP = np.zeros((n, 2*n))
+    W = np.zeros((1, 2*n))
+    for i in range(n):
+        SD = sqrtm(p)
+        SP[:, i] = (x + (math.sqrt(n) * SD[:, i]).reshape((n, 1))).flatten()
+        SP[:, i+n] = (x - (math.sqrt(n) * SD[:, i]).reshape((n, 1))).flatten()
+        W[:, i] = 1/(2*n)
+        W[:, i+n] = W[:, i]
+    return SP.astype(float), W.astype(float)
+
+
+# cubature kalman filter nonlinear prediction step
+def cubature_prediction(x_pred, p_pred):
+    n = np.shape(x_pred)[0]
+    [SP, W] = sigma(x_pred, p_pred)
+    x_pred = np.zeros((n, 1))
+    p_pred = q
+    for i in range(2*n):
+        x_pred = x_pred + (f(SP[:, i]).reshape((n, 1)) * W[0, i])
+    for i in range(2*n):
+        p_step = (f(SP[:, i]).reshape((n, 1)) - x_pred)
+        p_pred = p_pred + (p_step @ np.transpose(p_step) * W[0, i])
+    return x_pred.astype(float), p_pred.astype(float)
+
+
+# cubature kalman filter nonlinear update step
```

Comment:
Could you please use Python docstring format for function comment?
see: Python Docstrings (With Examples) https://www.programiz.com/python-programming/docstrings